### PR TITLE
CDAP-12438 add connector support for spark pipelines

### PR DESCRIPTION
**Pull request by albertshau**
Enhancing spark pipeline programs so that they can handle
connector plugins. They never used to have to worry about
connectors because they could do aggregations and joins in a
single spark program. But with the addition of conditions,
spark programs can have connectors as input and output as well.

As part of this, refactoring the ConnectorSource and ConnectorSink
so that there are two implementations. The first handles records
of different types and is used in mapreduce. The second is simpler
and handles writing just a single type of object and is used in
Spark, since Spark connectors are only used for conditions.

### CODE REVIEW

**Comment by albertshau** [comment]
https://builds.cask.co/browse/CDAP-RUT1283

File: cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/co/cask/cdap/etl/spark/function/PluginFunctionContext.java [hunk 1]
```diff
@@ -77,6 +81,15 @@ public PluginFunctionContext(StageSpec stageSpec, JavaSparkExecutionContext sec,
   }
 
   public <T> T createPlugin() throws Exception {
+    if (Constants.Connector.PLUGIN_TYPE.equals(stageSpec.getPluginType())) {
+      String connectorType = stageSpec.getPlugin().getProperties().get(Constants.Connector.TYPE);
+      // ok to pass in null to constructors here since we are only going to use the transform method
+      if (connectorType.equals(Constants.Connector.SOURCE_TYPE)) {
+        return (T) new SingleConnectorSource(null, null);
+      } else {
+        return (T) new SingleConnectorSink(null, null);
```
Comment (sagarkapare) [inline cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/co/cask/cdap/etl/spark/function/PluginFunctionContext.java:90]: no need to have it in `else` can be simply returned.

File: cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/co/cask/cdap/etl/spark/batch/ETLSpark.java [hunk 2]
```diff
@@ -139,17 +142,25 @@ public void initialize() throws Exception {
...
                                                                  phaseSpec.isProcessTimingEnabled());
+    PipelinePluginInstantiator pluginInstantiator =
+      new PipelinePluginInstantiator(pluginContext, context.getMetrics(), phaseSpec, new SingleConnectorFactory());
     final PipelineRuntime pipelineRuntime = new PipelineRuntime(context);
     final Admin admin = context.getAdmin();
 
-    for (final StageSpec stageSpec : phaseSpec.getPhase()) {
-      final String stageName = stageSpec.getName();
+    PipelinePhase phase = phaseSpec.getPhase();
+    // go through in topological order so that arguments set by one stage are seen be stages after it
```
Comment (sagarkapare) [inline cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/co/cask/cdap/etl/spark/batch/ETLSpark.java]: nit: typo `one stage are seen by stages`

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/batch/PipelinePluginInstantiator.java [hunk 3]
```diff
@@ -59,19 +64,47 @@ public PipelinePluginInstantiator(PluginContext pluginContext, Metrics metrics,
...
   public <T> T newPluginInstance(String stageName, MacroEvaluator macroEvaluator) throws InstantiationException {
+    T plugin = getBuiltIn(stageName);
+    if (plugin != null) {
+      return plugin;
+    }
+
+    return pluginContext.newPluginInstance(stageName, macroEvaluator);
+  }
+
+  private <T> T getBuiltIn(String stageName) {
```
Comment (sagarkapare) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/batch/PipelinePluginInstantiator.java]: nit: `@Nullable`?

**Review by sagarkapare — APPROVED** [review]
minor nits. LGTM 👍

**Pull request merged**